coderd/healthcheck/workspaceproxy.go

```diff
@@ -0,0 +1,137 @@
+package healthcheck
+
+import (
+	"context"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/buildinfo"
+	"github.com/coder/coder/v2/coderd/healthcheck/health"
+	"github.com/coder/coder/v2/coderd/util/ptr"
+	"github.com/coder/coder/v2/codersdk"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+type WorkspaceProxyReportOptions struct {
+	// CurrentVersion is the current server version.
+	// We pass this in to make it easier to test.
+	CurrentVersion string
+	// FetchWorkspaceProxies is a function that returns the available workspace proxies.
+	FetchWorkspaceProxies *func(context.Context) (codersdk.RegionsResponse[codersdk.WorkspaceProxy], error)
+	// UpdateProxyHealth is a function called when healthcheck is run.
+	// This would normally be ProxyHealth.ForceUpdate().
+	// We do this because if someone mashes the healthcheck refresh button
+	// they would expect up-to-date data.
+	UpdateProxyHealth *func(context.Context) error
+}
+
+// @typescript-generate WorkspaceProxyReport
+type WorkspaceProxyReport struct {
+	Healthy  bool            `json:"healthy"`
+	Severity health.Severity `json:"severity"`
+	Warnings []string        `json:"warnings"`
+	Error    *string         `json:"error"`
+
+	WorkspaceProxies codersdk.RegionsResponse[codersdk.WorkspaceProxy]
+}
+
+func (r *WorkspaceProxyReport) Run(ctx context.Context, opts *WorkspaceProxyReportOptions) {
+	r.Healthy = true
+	r.Severity = health.SeverityOK
+	r.Warnings = []string{}
+
+	if opts.FetchWorkspaceProxies == nil {
+		return
+	}
+	fetchWorkspaceProxiesFunc := *opts.FetchWorkspaceProxies
+
+	if opts.UpdateProxyHealth == nil {
+		err := "opts.UpdateProxyHealth must not be nil if opts.FetchWorkspaceProxies is not nil"
+		r.Error = ptr.Ref(err)
+		return
+	}
+	updateProxyHealthFunc := *opts.UpdateProxyHealth
+
+	// If this fails, just mark it as a warning. It is still updated in the background.
+	if err := updateProxyHealthFunc(ctx); err != nil {
+		r.Severity = health.SeverityWarning
+		r.Warnings = append(r.Warnings, xerrors.Errorf("update proxy health: %w", err).Error())
+		return
+	}
+
+	proxies, err := fetchWorkspaceProxiesFunc(ctx)
+	if err != nil {
+		r.Healthy = false
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref(err.Error())
+		return
+	}
+
+	r.WorkspaceProxies = proxies
+
+	var total, healthy int
+	for _, proxy := range r.WorkspaceProxies.Regions {
+		total++
+		if proxy.Healthy {
+			healthy++
+		}
+
+		if len(proxy.Status.Report.Errors) > 0 {
+			for _, err := range proxy.Status.Report.Errors {
+				r.appendError(xerrors.New(err))
+			}
+		}
+	}
+
+	r.Severity = calculateSeverity(total, healthy)
+	r.Healthy = r.Severity != health.SeverityError
+
+	// Versions _must_ match. Perform this check last. This will clobber any other severity.
+	for _, proxy := range r.WorkspaceProxies.Regions {
+		if vErr := checkVersion(proxy, opts.CurrentVersion); vErr != nil {
+			r.Healthy = false
+			r.Severity = health.SeverityError
```

Comment:
Not suggesting, just asking: Is this always an error? Could it be a warning instead if it's a patch version difference, for instance?

Reply:
Our current behaviour won't even let workspace proxies register if they don't match major.minor, but makes no further complaints if it's just a patch version difference.

Making it warn on a patch version difference sounds like a good idea in theory, but it will mean every time you upgrade, you'll get a short period of warning that there's a patch version diff. Although that's the current behaviour as well when you do a major version upgrade, so... 🤷  

I'll see about adding it in a follow-up.